macros/src/lib.rs

```diff
@@ -458,3 +459,141 @@ impl Parse for OptionSetter {
+    }
+
+    // Append utility fns to `impl` block item list
+    let opt_field_type = &opt_struct.ident;
+    impl_in.items.push(parse_quote! {
+        #[allow(unused)]
+        fn options(&mut self) -> &mut #opt_field_type {
+            self.options.get_or_insert_with(<#opt_field_type>::default)
+        }
+    });
+    impl_in.items.push(parse_quote! {
+        /// Set all options.  Note that this will replace all previous values set.
+        pub fn with_options(mut self, value: impl Into<Option<#opt_field_type>>) -> Self {
+            self.options = value.into();
+            self
+        }
+    });
+    // Append setter fns to `impl` block item list
+    for OptInfo { name, attrs, type_ } in opt_info {
+        let (accept, value) = if type_.is_ident("String")
+            || type_.is_ident("Bson")
+            || path_eq(&type_, &["bson", "Bson"])
+        {
+            (quote! { impl Into<#type_> }, quote! { value.into() })
+        } else if let Some(t) = inner_type(&type_, "Vec") {
+            (
+                quote! { impl IntoIterator<Item = #t> },
+                quote! { value.into_iter().collect() },
+            )
+        } else {
+            (quote! { #type_ }, quote! { value })
+        };
+        impl_in.items.push(parse_quote! {
+            #(#attrs)*
+            pub fn #name(mut self, value: #accept) -> Self {
+                self.options().#name = Some(#value);
+                self
+            }
+        })
+    }
+
+    // All done.
+    impl_in.to_token_stream().into()
+}
+
+#[import_tokens_attr]
+#[proc_macro_attribute]
+pub fn options_doc(
+    attr: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let setters = parse_macro_input!(attr as ItemImpl);
+    let mut impl_fn = parse_macro_input!(item as ImplItemFn);
+
+    // Collect a list of names from the setters impl
+    let mut setter_names = vec![];
+    for item in &setters.items {
+        match item {
+            ImplItem::Fn(item) if matches!(item.vis, Visibility::Public(..)) => {
+                setter_names.push(item.sig.ident.to_token_stream().to_string());
+            }
+            _ => continue,
+        }
+    }
+
+    // Get the rustdoc path to the action type, i.e. the type with generic arguments stripped
+    let mut doc_path = match &*setters.self_ty {
+        Type::Path(p) => p.path.clone(),
+        t => compile_error!(t.span(), "invalid options doc argument"),
+    };
+    for seg in &mut doc_path.segments {
+        seg.arguments = PathArguments::None;
+    }
+    let doc_path = doc_path.to_token_stream().to_string();
+
+    // Add the list of setters to the rustdoc for the fn
+    impl_fn.attrs.push(parse_quote! {
+        #[doc = ""]
+    });
+    impl_fn.attrs.push(parse_quote! {
+        #[doc = "These methods can be chained before calling `.await` to set options:"]
+    });
+    for name in setter_names {
+        let docstr = format!("  * [{0}]({1}::{0})", name, doc_path);
```

Comment:
nit: would it be possible to code format the method names (i.e. "`session`" instead of "session")? not sure if that's doable in a format string like this